constraints/recursiveCatEvals.js

```diff
@@ -53,29 +86,56 @@ var sCat = ["cp", "xp", "x0"];
  * ancestors before its maximality or minimality is used, your function will be
  * working with the correct values of isMin, isMax and parentCat.
  *
- * 7/29/19 refactor of an earlier version
+ * 10/26/20 update of an earlier version, now includes dummies.
  */
 
-function markMinMax(mytree){
+function markMinMax(mytree, options){
+	if(options.requireLexical){
+		mytree = createDummies(mytree, 'func');
+	}
+	if(options.requireOvertHead){
+		mytree = createDummies(mytree, 'silent');
+	}
+	return markMinMaxInner(mytree, options)
+}
+
+function markMinMaxInner(mytree, options){
 	/* If parentCat property is not already defined for this node, it is probably
 	 * the root node. Non-root nodes get the property parentCat when this node's
 	 * children are marked below.
 	 */
+	options = options || {};
+
 	if (!mytree.hasOwnProperty('parentCat')){
 		mytree.parentCat = "is root"; //marks the root node
 	}
 
-	//mark maximal nodes
-	mytree.isMax = (mytree.cat !== mytree.parentCat);
+	//Store the info of the most recent cat in order to skip over dummy nodes
+	//except when the dummy node's parent is ALSO a dummy node, then lastcat should be passed
+	//down dummy generation after dummy generation until a normal node is reached to inherit
+	//it as the parentCat.
+	if(mytree.cat === "dummy"){
+		mytree.isMax = false;
```

Comment:
I'm not sure there's any reason to label max/min on dummies, but I don't think it does any harm either.